ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -727,6 +732,53 @@ void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback)
       }
     }
 
+    @Override
+    public boolean check() {
+      if (quotaChargeCallback == null || isCharged) {
+        return true;
+      }
+      return quotaChargeCallback.check();
+    }
+
+    @Override
+    public boolean charge() {
+      if (quotaChargeCallback == null || isCharged) {
+        return true;
+      }
+      try {
+        quotaChargeCallback.charge(chunkSize);
```

Comment:
I think all these calls with eventually resolve to AmbryQuotaManager::charge(), however, that routine is not thread-safe.

Reply:
Per our chat IRL, you can handle them in a follow up PR before feature makes the prod.